src/ripple/app/misc/AmendmentTableImpl.cpp

```diff
@@ -569,19 +643,116 @@ AmendmentTableImpl::getJson (uint256 const& amendmentID)
+    void setMajorityTimesFromStateToDB (
+        std::vector<uint256> const& changedAmendments,
+        hash_map<uint256, AmendmentState>& amendmentMap) const;
+    ValidationSet getValidations (uint256 const& hash) const;
+};
+
+class AppApiFacadeMock final
+{
+public:
+    void setMajorityTimesFromDBToState (AmendmentState& toUpdate,
+                                        uint256 const& amendmentHash) const {};
+    void setMajorityTimesFromStateToDB (
+        std::vector<uint256> const& changedAmendments,
+        hash_map<uint256, AmendmentState>& amendmentMap) const {};
+    ValidationSet getValidations (uint256 const& hash) const
+    {
+        return ValidationSet ();
+    };
+};
+
+void AppApiFacadeImpl::setMajorityTimesFromDBToState (
+    AmendmentState& toUpdate,
+    uint256 const& amendmentHash) const
 {
-    return std::make_unique<AmendmentTableImpl> (majorityTime, majorityFraction,
-        journal);
+    std::string query =
+        "SELECT FirstMajority,LastMajority FROM Features WHERE hash='";
+    query.append (to_string (amendmentHash));
+    query.append ("';");
+
+    auto& walletDB (getApp ().getWalletDB ());
+    auto sl (walletDB.lock ());
+    auto db (walletDB.getDB ());
+
+    if (db->executeSQL (query) && db->startIterRows ())
+    {
+        toUpdate.m_firstMajority = db->getBigInt ("FirstMajority");
+        toUpdate.m_lastMajority = db->getBigInt ("LastMajority");
+        db->endIterRows ();
+    }
+}
+
+void AppApiFacadeImpl::setMajorityTimesFromStateToDB (
+    std::vector<uint256> const& changedAmendments,
+    hash_map<uint256, AmendmentState>& amendmentMap) const
+{
+    if (changedAmendments.empty ())
+        return;
+
+    auto& walletDB (getApp ().getWalletDB ());
+    auto sl (walletDB.lock ());
+    auto db (walletDB.getDB ());
+
+    db->executeSQL ("BEGIN TRANSACTION;");
+    for (auto const& hash : changedAmendments)
+    {
+        AmendmentState const& fState = amendmentMap[hash];
+        db->executeSQL (boost::str (boost::format (
+                                        "UPDATE Features SET FirstMajority "
+                                        "= %d WHERE Hash = '%s';") %
+                                    fState.m_firstMajority % to_string (hash)));
+        db->executeSQL (boost::str (boost::format (
+                                        "UPDATE Features SET LastMajority "
+                                        "= %d WHERE Hash = '%s';") %
+                                    fState.m_lastMajority % to_string (hash)));
+    }
+    db->executeSQL ("END TRANSACTION;");
+}
+
+ValidationSet AppApiFacadeImpl::getValidations (
+    uint256 const& hash) const
+{
+    return getApp ().getValidations ().getValidations (hash);
+}
+}  // AmendmentTableDetail
+
+std::unique_ptr<AmendmentTable> make_AmendmentTable (
+    std::chrono::seconds majorityTime,
+    int majorityFraction,
+    beast::Journal journal,
+    AmendmentTableDetail::AppApiFacade appApiFacade)
```

Comment:
I do not understand the need for all of this class enum rigamarole. First of all, callers would need to use declarations in a detail namespace in order to call this function, defeating the purpose. Second, the name of the enum matches the formal template argument name in the implementation class, a point of potential confusion. Third, factory functions like this should not contain business logic. Consider this declaration instead:
```
template <class Rep, class Period,
    class AppType = AmendmentTableDetail::AppApiFacadeImpl>>
std::unique_ptr<AmendmentTable>
make_AmendmentTable (std::chrono::duration<Rep, Period> const& majorityTime,
    int majorityFraction, beast::Journal journal);
```

Normal invocations use `AppApiFacadeImpl` while test code may use this alternate invocation:
```
...
make_AmendmentTable<AmendmentTableDetail::AppApiFacadeMock>(
    majorityTime, journal); 
...
```